Progress towards tier 1 support for FreeBSD x86_64

mgxm opened the pull request:
Continuing the work already done by @tiehuis and @myfreeweb, I have completed
some missing points to finish the Tier 1 support. One of the
main points was that we get rid of all the syscalls and now we are using the
libc interface, which is the stable ABI.

Almost all tests are passing, missing ~only one to fix and which I'm working
on~ a lot of fixes. The said test is the [test-cli](https://github.com/ziglang/zig/blob/master/test/cli.zig), which works on FreeBSD 12 but fails on
FreeBSD 11.

<details>
 <summary>FreeBSD 11 Stack Trace</summary>

```
error: NameTooLong
/usr/home/marcio/build/zig/build/lib/zig/std/os/index.zig:1040:28: 0x234bff in ??? (cli)
            posix.EPERM => return error.AccessDenied,
                           ^
/usr/home/marcio/build/zig/build/lib/zig/std/os/index.zig:1014:9: 0x22bc9c in ??? (cli)
        return deleteFileC(&file_path_c);
        ^
/usr/home/marcio/build/zig/build/lib/zig/std/os/index.zig:513:42: 0x2202f9 in ??? (cli)
    if (file_path.len >= posix.PATH_MAX) return error.NameTooLong;
                                         ^
/usr/home/marcio/build/zig/build/lib/zig/std/os/index.zig:1013:29: 0x22bc1b in ??? (cli)
        const file_path_c = try toPosixPath(file_path);
                            ^
/usr/home/marcio/build/zig/build/lib/zig/std/os/index.zig:1442:16: 0x229355 in ??? (cli)
            => return err,
               ^
/usr/home/marcio/build/zig/build/lib/zig/std/os/index.zig:1484:17: 0x2297ad in ??? (cli)
                try deleteTree(allocator, full_entry_path);
                ^
/usr/home/marcio/build/zig/test/cli.zig:40:9: 0x228411 in ??? (cli)
        try os.deleteTree(a, dir_path);
```
</details>

andrewrk commented:
Looks like the problem is in trying to cross compile, targeting FreeBSD.

```
lld: error: cannot open /usr/lib/Scrt1.o: No such file or directory
lld: error: cannot open /usr/lib/crti.o: No such file or directory
lld: error: cannot open /usr/lib/crtbegin.o: No such file or directory
lld: error: unable to find library -lgcc
lld: error: unable to find library -lgcc_s
lld: error: unable to find library -lc
lld: error: unable to find library -lm
lld: error: unable to find library -lgcc
lld: error: unable to find library -lgcc_s
lld: error: cannot open /usr/lib/crtend.o: No such file or directory
lld: error: cannot open /usr/lib/crtn.o: No such file or directory
```

We'll have to figure out how to target FreeBSD on non-FreeBSD operating systems to continue zig's promise of "use any target to build for any target".

However you don't have to solve this problem in this PR. One step at a time is OK, and this PR already does a lot.

mgxm commented:
> We'll have to figure out how to target FreeBSD on non-FreeBSD operating systems to continue zig's promise of "use any target to build for any target".

I completely forgot about that, I will work on it too

There's anything I can do?
should I remove Freebsd from the test matrix?

andrewrk commented:
> There's anything I can do?
> should I remove Freebsd from the test matrix?

Yeah, for now, remove it from the test matrix, and manually add the native tests in the CI script:
 * `bin/zig test ../test/behavior.zig`
 * `bin/zig test ../std/index.zig`
 * `bin/zig test ../std/special/compiler_rt/index.zig`

And then do these with --release-fast, --release-small, and --library c variants.

myfreeweb commented on std/os/index.zig:
```diff
@@ -116,7 +116,7 @@ pub fn getRandomBytes(buf: []u8) !void {
             }
         },
-        Os.macosx, Os.ios => return getRandomBytesDevURandom(buf),
+        Os.macosx, Os.ios, Os.freebsd => return getRandomBytesDevURandom(buf),
```
note: FreeBSD 12 has getrandom, I used it on purpose

andrewrk replied:
Why not use getrandom?

mgxm replied:
because getrandom just landed on FreeBSD 12 and I was considering we were going to support version 11, but as @myfreeweb  said if it's ok to no support, we can bring it back

andrewrk replied:
I see, thanks.

> I think it's fine to not support versions older than 12 in a young language.

I agree with this statement.

andrewrk commented:
> (btw do you target macOS from non-macOS operating systems?)

Yes, and it works, although LLD is not great at mach-o files. The newest darwin versions do not require the crt1.o startup files or anything like that.

> You can't cross-link without having the libraries.

We're going to have to provide these in source form (probably ported to zig, just like we do with compiler-rt) and have the ability to build them for any target.

myfreeweb commented:
> provide these in source form (probably ported to zig, just like we do with compiler-rt) and have the ability to build them for any target

A whole `libc` for any target? That sounds… very ambitious (and how does it align with "libc is the stable ABI"?)

andrewrk commented:
A whole libc for any target is #514. But that's not needed for FreeBSD cross compilation. Only crt1.o & friends, compatible with the FreeBSD libc is necessary. That's a much smaller component. These .o files are statically linked into executables when building against FreeBSD libc. Try it for yourself with the system compiler and look at the linker line. So your ABI question would apply to using C with the system compiler as much as it would apply to Zig here.

mgxm commented:
I removed the FreeBSD from the test matrix and added the native's tests as
recommended. For this last one, please tell me if it's the correct way or I got
it wrong

mgxm commented:
test ../std/index.zig will fail, clock-related functions need to be implemented yet.

andrewrk commented:
That's ok, let's disable that test for this PR and then part of tier 1 support is working on those tests.

andrewrk commented:
I'm ready to merge this, do you consider it ready @mgxm?

mgxm commented:
Yes, but because I made a mess running the tests in the beginning, I think we
should rename the PR to "Initial Tier 1 support" or something like that. I was
using all the tests running with `./bin/zig build --build-file ../build.zig
test` as a reference and these tests are all passing on FreeBSD 12 as I didn't
know about and didn't go through the tests with `bin/zig test ../std/index.zig`,
I could not verify all the missing fixes, so the cross-compilation is not yet
done and the std library has some work to be done.

My apologies for the confusion.

mgxm commented:
Issue #1759 have a checklist that you could ticked off.

- switch to libc
- stack traces

The issue #1217 I could not reproduce on the master branch

andrewrk commented:
No worries! This patch still represents significant progress. Thank you for your contribution.
